c2rust-analyze/src/main.rs

```diff
@@ -150,6 +150,18 @@ fn run(tcx: TyCtxt) {
         gacx.fn_sigs.insert(ldid.to_def_id(), lsig);
     }
 
+    // Label the field types of each struct.
+    for ldid in tcx.hir_crate_items(()).definitions() {
+        let did = ldid.to_def_id();
+        if tcx.def_kind(did) != DefKind::Struct {
+            continue;
+        }
+        for field in tcx.adt_def(did).all_fields() {
+            let lty = gacx.assign_pointer_ids(tcx.type_of(field.did));
+            gacx.field_tys.insert(field.did, lty);
+        }
+    }
```

Comment:
Could do this (just a suggestion):

```suggestion
    gacx.field_tys = tcx
        .hir_crate_items(())
        .definitions()
        .map(|ldid| ldid.to_def_id())
        .filter(|did| tcx.def_kind(did) == DefKind::Struct)
        .flat_map(|did| tcx.adt_def(did).all_fields())
        .map(|field| {
            let lty = gacx.assign_pointer_ids(tcx.type_of(field.did));
            (field.did, lty)
        })
        .collect();
```

Reply:
the parts of me that like functional programming like this, but i feel like the way it is now is more scrutable